### PR TITLE
integration test + infrastructure for testing in task-coordinator (demo purposes)

### PR DESCRIPTION
**Pull request by AfanasievYuriy**

### CODE REVIEW

File: taskcoordinator/app/build.gradle [hunk 1]
```diff
@@ -35,4 +35,5 @@ dependencies {
     compileOnly 'org.projectlombok:lombok:1.18.4'
     annotationProcessor 'org.projectlombok:lombok:1.18.4'
     testCompile 'junit:junit:4.12'
+    testCompile 'org.springframework.boot:spring-boot-starter-test:2.1.1.RELEASE'
```
Comment (andrsuh) [inline taskcoordinator/app/build.gradle]: remove version. it is handled by dependency plugin

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDto.java [hunk 2]
```diff
@@ -21,9 +21,13 @@
...
 
     public LaundryStateDto(LaundryStateEntity laundryStateEntity) {
-        this.laundryId = laundryStateEntity.getId();
+        this.id = laundryStateEntity.getId();
         this.queueWaitingTime = laundryStateEntity.getQueueWaitingTime();
         this.reservedTime = laundryStateEntity.getReservedTime();
         this.version = laundryStateEntity.getVersion();
     }
+
+    public long getCompletionTimePrediction() {
```
Comment (andrsuh) [inline taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDto.java]: It is more then weird to place prediction logic in dto.

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java [hunk 3]
```diff
@@ -23,11 +23,23 @@
 
     private OrderService orderService;
 
+    @Override
+    public LaundryStateDto getLaundryStateById(int laundryId) {
+        if (laundryId < 0 ) {
+            throw new IllegalArgumentException("id or orderDuration can't be < 0");
```
Comment (andrsuh) [inline taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java]: change message

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDto.java [hunk 4]
```diff
@@ -21,9 +21,13 @@
     private Integer version;
 
     public LaundryStateDto(LaundryStateEntity laundryStateEntity) {
-        this.laundryId = laundryStateEntity.getId();
```
Comment (andrsuh) [inline taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDto.java:24]: null might be passed

File: taskcoordinator/app/src/test/java/com/microservices/taskcoordinator/controllers/CoordinatorControllerIntegrationTests.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+    @Autowired
+    private OrderService orderService;
+
+    private ObjectMapper objectMapper = new ObjectMapper();
+
+    @Test
+    @Sql(scripts = "/test-data/basic.sql")
+    @Transactional
+    public void t() throws Exception {
```
Comment (andrsuh) [inline taskcoordinator/app/src/test/java/com/microservices/taskcoordinator/controllers/CoordinatorControllerIntegrationTests.java]: name :)

File: taskcoordinator/app/src/test/java/com/microservices/taskcoordinator/controllers/CoordinatorControllerIntegrationTests.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+        LaundryStateDto laundryWillBeChosen = laundryStateService.getLeastLoadedLaundry();
+
+        List<OrderDetailDto> inputOrderDetails = new ArrayList<>(Arrays.asList(
+                new OrderDetailDto(10, DEFAULT_DETAIL_WEIGHT, DEFAULT_DETAIL_DURATION, INPUT_ORDER_ID),
+                new OrderDetailDto(11, DEFAULT_DETAIL_WEIGHT, DEFAULT_DETAIL_DURATION, INPUT_ORDER_ID)
+        ));
+        OrderCoordinationDto coordinationDto = new OrderCoordinationDto(INPUT_ORDER_ID, inputOrderDetails);
+
+        String inputRoleDtoJson = objectMapper.writeValueAsString(coordinationDto);
```
Comment (andrsuh) [inline taskcoordinator/app/src/test/java/com/microservices/taskcoordinator/controllers/CoordinatorControllerIntegrationTests.java]: name

File: taskcoordinator/app/src/test/java/com/microservices/taskcoordinator/controllers/CoordinatorControllerIntegrationTests.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        String responseOrderSubmission = mockMvc.perform(MockMvcRequestBuilders.post("/orders")
+                .contentType(APPLICATION_JSON.toString())
+                .content(inputRoleDtoJson))
+                .andExpect(status().isOk())
+                .andReturn().getResponse().getContentAsString();
+        OrderSubmissionDto orderSubmissionDto = objectMapper.readValue(responseOrderSubmission, OrderSubmissionDto.class);
+
+        //checkout order and laundryState form DB to see a difference
```
Comment (andrsuh) [inline taskcoordinator/app/src/test/java/com/microservices/taskcoordinator/controllers/CoordinatorControllerIntegrationTests.java]: from

File: taskcoordinator/app/src/test/resources/application.properties [hunk 8]
```diff
@@ -0,0 +1,12 @@
+database.schema.name=task_coordinator
```
Comment (andrsuh) [inline taskcoordinator/app/src/test/resources/application.properties:1]: is this needed?

**Review by andrsuh — COMMENTED** [review]
Looks pretty cool, just a few comments!

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDto.java (hunk 2 above)
Comment (AfanasievYuriy) [reply taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDto.java]: yeah, i agree, but where should i move it? it's not convenient to have it entity (for now), cause i work with dtos mostly
hm

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java (hunk 3 above)
Comment (AfanasievYuriy) [reply taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java]: ))

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDto.java (hunk 4 above)
Comment (AfanasievYuriy) [reply taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDto.java:24]: so it should be reworked in all constructors like dto-entity/entity-dto, right?

File: taskcoordinator/app/src/test/java/com/microservices/taskcoordinator/controllers/CoordinatorControllerIntegrationTests.java (hunk 5 above)
Comment (AfanasievYuriy) [reply taskcoordinator/app/src/test/java/com/microservices/taskcoordinator/controllers/CoordinatorControllerIntegrationTests.java]: double Ѣ

File: taskcoordinator/app/src/test/java/com/microservices/taskcoordinator/controllers/CoordinatorControllerIntegrationTests.java (hunk 6 above)
Comment (AfanasievYuriy) [reply taskcoordinator/app/src/test/java/com/microservices/taskcoordinator/controllers/CoordinatorControllerIntegrationTests.java]: ?

File: taskcoordinator/app/src/test/resources/application.properties (hunk 8 above)
Comment (AfanasievYuriy) [reply taskcoordinator/app/src/test/resources/application.properties:1]: I'll explain it all in person

File: taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java [hunk 9]
```diff
@@ -81,8 +93,8 @@ public LaundryStateDto updateLaundryStateWithOrderProcessed(OrderProcessedDto or
 
     @Override
     @Transactional
-    public LaundryStateEntity getLeastLoadedLaundry() {
-        return laundryStateRepository.getLeastLoadedLaundry();
+    public LaundryStateDto getLeastLoadedLaundry() {
+        return new LaundryStateDto(laundryStateRepository.getLeastLoadedLaundry());
```
Comment (andrsuh) [inline taskcoordinator/app/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java]: make it return Optional

**Review by andrsuh — APPROVED** [review]

**Pull request merged**